### PR TITLE
resource/aws_eip_association: fix eventual consistency issue when associating EIP

### PR DESCRIPTION
**Pull request by dohoangkhiem**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates #11617
Relates #12449

### Terraform Version
```
$ terraform -v
Terraform v0.12.16
+ provider.aws v2.39.0
```

### Affected Resource(s)

aws_eip_association

### Terraform Configuration
```
provider "aws" {
  version = "2.39.0"
  region = var.aws_region
}

resource "aws_instance" "this" {
  ami = "ami-06fb5332e8e3e577a"
  instance_type = "t2.nano"
  key_name = "default_key"
  vpc_security_group_ids = ["sg-923d3cf4"]
  subnet_id = "subnet-61a23e06"
  associate_public_ip_address = true

  root_block_device {
    volume_size = 20
    volume_type = "gp2"
    delete_on_termination = true
  }

  lifecycle {
    create_before_destroy = true
  }
}

resource "aws_eip" "this" {
  vpc = true
}

resource "aws_eip_association" "this" {
  instance_id   = aws_instance.this.id
  allocation_id = aws_eip.this.id
}
```

### Expected Behavior

`aws_eip_association` is created successfully in TF state and in EC2.

### Actual Behavior
Rarely, the EIP is associated, but in TF state the resource is not recorded due to eventual consistency issue when reading back the association

```
Error: Provider produced inconsistent result after apply
When applying changes to aws_eip_association.this, provider "aws" produced an unexpected new value for was present, but now absent. This is a bug in the provider, which should be reported in the provider's own issue tracker.
``` 

Output from acceptance testing:

```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSEIPAssociation_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSEIPAssociation_ -timeout 120m
=== RUN   TestAccAWSEIPAssociation_instance
=== PAUSE TestAccAWSEIPAssociation_instance
=== RUN   TestAccAWSEIPAssociation_networkInterface
=== PAUSE TestAccAWSEIPAssociation_networkInterface
=== RUN   TestAccAWSEIPAssociation_basic
=== PAUSE TestAccAWSEIPAssociation_basic
=== RUN   TestAccAWSEIPAssociation_ec2Classic
=== PAUSE TestAccAWSEIPAssociation_ec2Classic
=== RUN   TestAccAWSEIPAssociation_spotInstance
=== PAUSE TestAccAWSEIPAssociation_spotInstance
=== RUN   TestAccAWSEIPAssociation_disappears
=== PAUSE TestAccAWSEIPAssociation_disappears
=== CONT  TestAccAWSEIPAssociation_instance
=== CONT  TestAccAWSEIPAssociation_spotInstance
=== CONT  TestAccAWSEIPAssociation_disappears
=== CONT  TestAccAWSEIPAssociation_ec2Classic
=== CONT  TestAccAWSEIPAssociation_networkInterface
=== CONT  TestAccAWSEIPAssociation_basic
=== CONT  TestAccAWSEIPAssociation_ec2Classic
    ec2_classic_test.go:54: this test can only run in EC2-Classic, platforms available in us-east-1: ["VPC"]
--- SKIP: TestAccAWSEIPAssociation_ec2Classic (5.79s)
--- PASS: TestAccAWSEIPAssociation_networkInterface (75.35s)
--- PASS: TestAccAWSEIPAssociation_instance (108.61s)
--- PASS: TestAccAWSEIPAssociation_disappears (109.13s)
--- PASS: TestAccAWSEIPAssociation_basic (169.76s)
--- PASS: TestAccAWSEIPAssociation_spotInstance (310.48s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	310.577s
```

### CODE REVIEW

File: aws/resource_aws_eip_association.go [hunk 1]
```diff
@@ -157,7 +157,30 @@ func resourceAwsEipAssociationRead(d *schema.ResourceData, meta interface{}) err
 		return err
 	}
 
-	response, err := conn.DescribeAddresses(request)
+	var response *ec2.DescribeAddressesOutput
+	err = resource.Retry(2*time.Minute, func() *resource.RetryError {
+		var err error
+		response, err = conn.DescribeAddresses(request)
+
+		if tfawserr.ErrCodeEquals(err, "InvalidAssociationID.NotFound") {
```
Comment (bflad) [inline aws/resource_aws_eip_association.go:165]: It seems like we should also only retry this error if it is during resource recreation:

```suggestion
		if d.IsNewResource() && tfawserr.ErrCodeEquals(err, "InvalidAssociationID.NotFound") {
```

File: aws/resource_aws_eip_association.go [hunk 2]
```diff
@@ -157,7 +157,30 @@ func resourceAwsEipAssociationRead(d *schema.ResourceData, meta interface{}) err
...
+		if err != nil {
+			return resource.NonRetryableError(err)
+		}
+
+		return nil
+	})
+
+	if isResourceTimeoutError(err) {
+		response, err = conn.DescribeAddresses(request)
+	}
```
Comment (bflad) [inline aws/resource_aws_eip_association.go:182]: Before returning it as an error below, it seems like we should also remove from state to trigger recreation if it is not a new resource.

```suggestion
	if isResourceTimeoutError(err) {
		response, err = conn.DescribeAddresses(request)
	}

	if !d.IsNewResource() && tfawserr.ErrCodeEquals(err, "InvalidAssociationID.NotFound") {
		log.Printf("[WARN] EIP Association (%s) not found, removing from state", d.Id())
		d.SetId("")
		return nil
	}
```

File: aws/resource_aws_eip_association.go [hunk 3]
```diff
@@ -157,7 +157,30 @@ func resourceAwsEipAssociationRead(d *schema.ResourceData, meta interface{}) err
 		return err
 	}
 
-	response, err := conn.DescribeAddresses(request)
+	var response *ec2.DescribeAddressesOutput
+	err = resource.Retry(2*time.Minute, func() *resource.RetryError {
```
Comment (bflad) [inline aws/resource_aws_eip_association.go:161]: Nit: We should create a constant for this time value (check out `aws/internal/service/ec2/waiter.go`). See also: https://github.com/hashicorp/terraform-provider-aws/issues/16753

**Review by bflad — CHANGES_REQUESTED** [review]
Hi @dohoangkhiem 👋  Thank you for submitting this fix. It is definitely on the right track -- just a few minor things and should be good to go. 👍

File: aws/resource_aws_eip_association.go (hunk 3 above)
Comment (dohoangkhiem) [reply aws/resource_aws_eip_association.go]: Yes, I added a general PropagationTimeout for EC2-related resources there (as of now there's only `InstanceAttributePropagationTimeout`)

File: aws/resource_aws_eip_association.go (hunk 1 above)
Comment (dohoangkhiem) [reply aws/resource_aws_eip_association.go]: true, updated. Thanks @bflad !

**Comment by dohoangkhiem** [comment]
Many thanks @bflad for suggestions, I updated PR, acc tests passed.

**Review by bflad — APPROVED** [review]
Looks great, thank you @dohoangkhiem 🚀 

Output from acceptance testing in AWS Commercial:

```
--- PASS: TestAccAWSEIPAssociation_networkInterface (67.84s)
--- PASS: TestAccAWSEIPAssociation_ec2Classic (82.56s)
--- PASS: TestAccAWSEIPAssociation_instance (95.90s)
--- PASS: TestAccAWSEIPAssociation_disappears (103.23s)
--- PASS: TestAccAWSEIPAssociation_basic (170.68s)
--- PASS: TestAccAWSEIPAssociation_spotInstance (302.76s)
```

Output from acceptance testing in AWS GovCloud (US):

```
--- PASS: TestAccAWSEIPAssociation_networkInterface (71.88s)
--- PASS: TestAccAWSEIPAssociation_instance (97.48s)
--- PASS: TestAccAWSEIPAssociation_disappears (104.81s)
--- PASS: TestAccAWSEIPAssociation_basic (181.36s)
--- PASS: TestAccAWSEIPAssociation_spotInstance (337.15s)
--- SKIP: TestAccAWSEIPAssociation_ec2Classic (2.41s)
```

**Pull request merged**